### PR TITLE
[POC] Behat context modifiers

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This is POC that aims to provide a simple mechanism that would allow us to run our behat tests in multiple configurations: different languages, currencies, settings ...
| Type?         | improvement
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? |
| How to test?  | 

## Background

We have now a nice behat test suite that covers lot of scenarios, including cart prices computations. However this suite is run for one currency, one language. We would like to run it for all configurations without having to copy-paste ou scenarios, in order to keep them maintenable.

## Idea

My idea is to introduce into key FeatureConcepts some flags that control settings, and have Modifiers that can switch these flags on and off. Then in order to run the suite with a specific flag, you run the __right behat configuration file which includes and loads this Modifier as a Context__ (see example https://github.com/PrestaShop/PrestaShop/pull/12987/files#diff-9eb43ef5fdf61a7763f278f9e142aef7).

So if you want to run the suite with 1,000 different currencies, you need 1,000 behat config files. But gherkin scenarios are not modified.

## The POC

This POC introduces 1 flag, the related Modifier and the related config file. It allows to modify the ProductFeatureContext behavior and uses virtual products instead of real ones.

## Want to test ?

Pull this branch, and run
`php -d date.timezone=UTC ./vendor/bin/behat -c tests/Integration/Behaviour/behat-virtual-product-modifier.yml`
instead of the regular
`php -d date.timezone=UTC ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/12987)

### CODE REVIEW

File: src/PrestaShopBundle/Form/Admin/AdvancedParameters/Performance/PerformanceFormHandler.php [hunk 1]
```diff
@@ -34,6 +34,8 @@
 /**
  * This class manages the data manipulated using forms
  * in "Configure > Advanced Parameters > Performance" page.
+ *
+ * @deprecated since 1.7.4.0, to be removed in 1.8
```
Comment (matks) [inline src/PrestaShopBundle/Form/Admin/AdvancedParameters/Performance/PerformanceFormHandler.php:38]: Mistake, please ignore

File: tests/Integration/Behaviour/behat-virtual-product-modifier.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+                - Tests\Integration\Behaviour\Features\Context\CartFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\ProductFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\CartRuleFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\CurrencyFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\TaxFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\SpecificPriceRuleFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\CarrierFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\CustomerFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\OrderFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\Modifier\ProductFeatureContextModifier
```
Comment (matks) [inline tests/Integration/Behaviour/behat-virtual-product-modifier.yml:30]: This Modifier here is the thing that matters

**Comment by matks** [comment]
Note: I thought about 4 solutions to handle this need

1) Override FeatureContext to modify `protected` functions, however since registering FeatureContext for behat also applies autoload, replacing a FeatureContext by its child crashes it

2) The Modifiers solution implemented here, but it requires 1 dedicated config file for each settings (note: we might be able to generate them). Also YAML seems to have inheritance capabilities so maybe we can avoid a lot of copy-paste.

3) Using environment variables for behat, so we would have something like `vendor/bin/behat -c behat.yml --env={foo: 'bar'}` . To run different settings, we would consequently do
```vendor/bin/behat -c behat.yml --env={lang: 'FR', currency: 'EUR'}
vendor/bin/behat -c behat.yml --env={lang: 'EN', currency: 'USD'}
vendor/bin/behat -c behat.yml --env={lang: 'NG', currency: 'AZD'}
```

4) There's also behat profiles
http://behat.org/en/latest/user_guide/configuration.html#overriding-default-params
but that would require a very huge behat config file with all profiles in it.

**Comment by tomlev** [comment]
@matks I like the concept

But I think we must flag these scenarios with more than what you propose. Consider the following usecase:
I want to rerun for modified products an entire directory, EXCEPT some tests (eg carrier specific fees, which should be null on virtual prodcuts)

So maybe a good way could be using your context modifier concept, but adding flags to run only some scenarios.
With that, process will be:
- I create the modifier context to add some behaviour and rerun scenarios
- I flag each scenario I want to rerun with modified behaviour

wdyt ?

**Comment by matks** [comment]
> @matks I like the concept
> 
> But I think we must flag these scenarios with more than what you propose. Consider the following usecase:
> I want to rerun for modified products an entire directory, EXCEPT some tests (eg carrier specific fees, which should be null on virtual prodcuts)
> 
> So maybe a good way could be using your context modifier concept, but adding flags to run only some scenarios.
> With that, process will be:
> 
> * I create the modifier context to add some behaviour and rerun scenarios
> * I flag each scenario I want to rerun with modified behaviour
> 
> wdyt ?

Sorry for the late answer.

Idea looks good 👍but we need to find the way to flag the scenarios "suitable for modified run".

If we use behat tags, I'm afraid we'll end up with some dirty-looking scenarios such as
```
@ok-for-virtual-products
@ok-for-euro
@nok-for-multishop
@nok-for-whatever
Scenario: 4 products in cart, one is virtual, several quantities, one 5€ global voucher
    Given I have an empty default cart
    Given shop configuration for "PS_CART_RULE_FEATURE_ACTIVE" is set to 1
    Given there is a product in the catalog named "product1" with a price of 19.812 and 1000 items in stock
    Given there is a product in the catalog named "product2" with a price of 32.388 and 1000 items in stock
    Given there is a product in the catalog named "product3" with a price of 31.188 and 1000 items in stock
    Given there is a product in the catalog named "product8" with a price of 12.345 and 1000 items in stock
```

What about a `Given` step that would stop the execution is given settings are not compatible with the tests ?
```
Given this test is not compatible with virtual products (because ...)
```
Using regexp we can even put in () the reason why it is not suitable for situation X.

This works if we have a blacklist of scenarios rather than a whitelist of scenarios. For example if we have 200 scenarios and only 20 of them cannot be run in virtual products usecase, it's OK. If it's rather 180, we're going to pollute 180 tests with a Step 🤔...

**Comment by tomlev** [comment]
we must use whitelist and not blacklist here:
modified scenarios must be defined by feature, else we'll need to change all the existing scenarios to blacklist them

so the following example is good to me:

```
@test-with-modifier-virtual-products
@test-with-modifier-multishop
Scenario: ...
```

in that case we will see changes on scenarios when they are processed with modifier

**Comment by matks** [comment]
Closed in favor of https://github.com/PrestaShop/PrestaShop/pull/13596

**Pull request closed without merging**